File: js-compressor/lz.js
function compress(data) {
    const result = [];
    const windowSize = 20;

    let i = 0;
    while (i < data.length) {
        let matchOffset = 0;
        let matchLength = 0;

        const start = Math.max(0, i - windowSize);

        for (let j = start; j < i; j++) {
            let k = 0;
            while (
                i + k < data.length &&
                data[j + k] === data[i + k] &&
                k < 255 &&
                j + k < i
            ) {
                k++;
            }

            if (k >= 3 && k > matchLength) {
                matchOffset = i - j;
                matchLength = k;
            }
        }

        if (matchLength >= 3) {
            result.push(0x01, matchOffset, matchLength);
            i += matchLength;
        } else {
            result.push(0x00, data[i]);
            i++;
        }
    }

    return Buffer.from(result);
}

function decompress(data) {
    const result = [];

    let i = 0;
    while (i < data.length) {
        const flag = data[i];
        if (flag === 0x00) {
            result.push(data[i + 1]);
            i += 2;
        } else if (flag === 0x01) {
            const offset = data[i + 1];
            const length = data[i + 2];
            const start = result.length - offset;
            for (let j = 0; j < length; j++) {
                result.push(result[start + j]);
            }
            i += 3;
        } else {
            throw new Error("Invalid LZ token");
        }
    }

    return Buffer.from(result);
}

module.exports = { compress, decompress };

File: rust-compressor/src/lz.rs
pub fn compress_lz(data: &[u8]) -> Vec<u8> {
    let mut result = Vec::new();
    let window_size = 20;
    let mut i = 0;

    while i < data.len() {
        let mut match_offset = 0;
        let mut match_length = 0;

        let start = if i > window_size { i - window_size } else { 0 };

        for j in start..i {
            let mut k = 0;
            while i + k < data.len()
                && data[j + k] == data[i + k]
                && k < 255
                && j + k < i
            {
                k += 1;
            }

            if k > match_length && k >= 3 {
                match_offset = i - j;
                match_length = k;
            }
        }

        if match_length >= 3 {
            result.push(0x01);
            result.push(match_offset as u8);
            result.push(match_length as u8);
            i += match_length;
        } else {
            result.push(0x00);
            result.push(data[i]);
            i += 1;
        }
    }

    result
}

pub fn decompress_lz(data: &[u8]) -> Vec<u8> {
    let mut result = Vec::new();
    let mut i = 0;

    while i < data.len() {
        let flag = data[i];
        if flag == 0x00 {
            result.push(data[i + 1]);
            i += 2;
        } else if flag == 0x01 {
            let offset = data[i + 1] as usize;
            let length = data[i + 2] as usize;
            let start = result.len() - offset;

            for j in 0..length {
                result.push(result[start + j]);
            }

            i += 3;
        } else {
            panic!("Invalid LZ77 flag: {}", flag);
        }
    }

    result
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_lz_roundtrip() {
        let input = b"ABABABABABAB";
        let compressed = compress_lz(input);
        let decompressed = decompress_lz(&compressed);
        assert_eq!(input.to_vec(), decompressed);
    }
}
